### DISCUSSION TITLE
Filter files with arbitrary Lua function

### DISCUSSION THREAD

**Discussion by dxrcy** [nvim-tree/nvim-tree.lua#2648] [Ideas]
Is it possible to add support for a custom filter function, so files can be filtered by means other than the filename matching a pattern. An example use for this is hiding executables with no file extension.

An example implementation using a non-existant shell command `is-binary`, which echoes `true` if the file is a binary executable such as ELF.

```lua
require("nvim-tree").setup {
    filters = {
        custom2 = function(filename)
            return vim.fn.system("is-binary " .. filename) == "true"
        end
    }
}
```

**Comment by alex-courtis** [comment]
There is precedent: `:h nvim-tree-opts-sort`

#2616 would add more fields, we could include `uv.uv_fs_t`

Please raise a feature request. [Pull Requests](https://github.com/nvim-tree/nvim-tree.lua/blob/master/CONTRIBUTING.md) are most gratefully appreciated.

**Reply by dxrcy** [reply]
Would a custom function, or a new field (on `nvim-tree-opts-filters` or `nvim-tree-opts-sort`) be preferable?

**Reply by alex-courtis** [reply]
We can overload `filters.custom` to take a string table or function, similar to `renderer.root_folder_label`.

`lib.clone_node` can provide the sanitized nodes to pass to the user's function.
